plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Tokenize.hs

```diff
@@ -0,0 +1,202 @@
+import           Language.LSP.Protocol.Types     (Position (Position),
+                                                  Range (Range), UInt, mkRange)
+import           Language.LSP.VFS                hiding (line)
+import           Prelude                         hiding (length, span)
+
+type Tokenizer m a = forall t. StateT (PTokenState t) m a
+
+
+data PTokenState t = PTokenState
+  { rangeIdSetMap       :: RangeIdSetMap,
+    rope                :: Rope,
+    cursor              :: Char.Position,
+    columnsInUtf16      :: UInt,
+    currentRange        :: Range,
+    currentRangeContext :: SplitResult
+  }
+
+runTokenizer :: (Monad m) => Tokenizer m a -> PTokenState t -> m RangeIdSetMap
+runTokenizer p st = rangeIdSetMap <$> execStateT p st
+
+data SplitResult
+  = NoSplit (Text, Range) -- does not need to split, token text, token range
+  | Split (Text, Range, Range) -- token text, prefix range(module range), token range
+  deriving (Show)
+
+startRange :: Range
+startRange = Range (Position 0 0) (Position 0 0)
+
+mkPTokenState :: VirtualFile -> PTokenState a
+mkPTokenState vf =
+  PTokenState
+    { rangeIdSetMap = mempty,
+      rope = Rope.fromText $ toText vf._file_text,
+      cursor = Char.Position 0 0,
+      columnsInUtf16 = 0,
+      currentRange = startRange,
+      currentRangeContext = NoSplit ("", startRange)
+    }
+
+addRangeIdSetMap :: (Monad m) => Range -> Identifier -> Tokenizer m ()
+addRangeIdSetMap r i = modify $ \s -> s {rangeIdSetMap = Map.insertWith (<>) r (S.singleton i) $ rangeIdSetMap s}
+
+-- lift a Tokenizer Maybe () to Tokenizer m (),
+-- if the Maybe is Nothing, do nothing, recover the state
+-- if the Maybe is Just (), do the action, and keep the state
+liftMaybeM :: (Monad m) => Tokenizer Maybe () -> Tokenizer m ()
+liftMaybeM p = do
+  st <- get
+  forM_ (execStateT p st) put
+
+hieAstSpanIdentifiers :: VirtualFile -> HieAST a -> RangeIdSetMap
+hieAstSpanIdentifiers vf ast = runIdentity $ runTokenizer (foldAst ast) (mkPTokenState vf)
+
+-- | foldAst
+-- visit every leaf node in the ast in depth first order
+foldAst :: (Monad m) => HieAST t -> Tokenizer m ()
+foldAst ast = if null (nodeChildren ast)
+  then liftMaybeM (visitLeafIds ast)
+  else mapM_ foldAst $ nodeChildren ast
+
+visitLeafIds :: HieAST t -> Tokenizer Maybe ()
+visitLeafIds leaf = liftMaybeM $ do
+  let span = nodeSpan leaf
+  (ran, token) <- focusTokenAt leaf
+  -- if `focusTokenAt` succeed, we can safely assume we have shift the cursor correctly
+  -- we do not need to recover the cursor state, even if the following computation failed
+  liftMaybeM $ do
+    -- only handle the leaf node with single column token
+    guard $ srcSpanStartLine span == srcSpanEndLine span
+    splitResult <-  lift $ splitRangeByText token ran
+    modify $ \s -> s {currentRange = ran, currentRangeContext = splitResult}
+    mapM_ combineNodeIds $ Map.filterWithKey (\k _ -> k == SourceInfo) $ getSourcedNodeInfo $ sourcedNodeInfo leaf
```

Comment:
Could we remove `currentRange` and `currentRangeContext` from the state if we passed as arguments to `combineNodeIds`?